ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java

```diff
@@ -385,25 +408,25 @@ public void setPivotBlockSelector(final DynamicPivotBlockSelector pivotBlockSele
     this.pivotBlockSelector = pivotBlockSelector;
   }
 
-  public BlockAddedObserver getBlockAddedListener() {
+  public BlockAddedObserver createBlockchainObserver() {
     return addedBlockContext -> {
-      if (snapSyncState.isWaitingBlockchain()) {
-        // if we receive a new pivot block we can restart the heal
-        pivotBlockSelector.check(
-            (____, isNewPivotBlock) -> {
-              if (isNewPivotBlock) {
-                snapSyncState.setWaitingBlockchain(false);
-              }
-            });
-        // if we are close to the head we can also restart the heal and finish snapsync
-        if (!pivotBlockSelector.isBlockchainBehind()) {
-          snapSyncState.setWaitingBlockchain(false);
-        }
-        if (!snapSyncState.isWaitingBlockchain()) {
-          blockObserverId.ifPresent(blockchain::removeObserver);
-          blockObserverId = OptionalLong.empty();
-          reloadTrieHeal();
-        }
+      final AtomicBoolean foundNewPivotBlock = new AtomicBoolean(false);
+      pivotBlockSelector.check(
```

Comment:
is it not a problem that `check` executes async code when it determines that the pivot block is too far behind ?  For example could we incorrectly return no new pivot if we have no peers.  

The async code in DynamicPivotBlockSelector makes this hard to reason about

Reply:
Indeed, the code for the Pivot Block Selector is a bit complex to understand and could benefit from more detailed descriptions.

This class fetches pivot blocks in advance to avoid having to search at the time of the call. I don't have the exact numbers, but for example:

I call the check method and if the distance to the head is >60 and <126, I will cache a pivot block. I won't use it right now. On the next call, when the distance to the head is close to 126, I will take the one cached. This means that every time we call this method and we are in this range(>60 and <128), we are looking for a pivot block to put in a cache ,  This is to have fast access to the next pivot block

In conclusion, there will be no problem as it will be synchronous (every time we will take the pivot from cache).

Reply:
it would probably be beneficial to simplify this pivot block selection code as it's more technical debt. Previously, obtaining a pivot block required querying 5 peers and reaching a consensus, which could be time-consuming. Now, it's simply a matter of taking the last safe block. Therefore, this logic may no longer be necessary.